added new method get_nth_day_of_week

mahdix opened the pull request.

Commit:
added new method get_nth_day_of_week

Commit:
updating version

mm-binary commented:
Your suggestion makes sense, Matt.

Commit:
now using day_of_week names instead of numbers

mwmiller commented on t/object_methods.t:
```diff
@@ -85,3 +85,6 @@ is($datetime3->minus_time_interval('1d')->is_same_as($datetime2),  1,          '
 is($datetime2->minus_time_interval('-1d')->is_same_as($datetime3), 1,          'minus_time_interval("-1d") yields one day ahead.');
 throws_ok { $datetime3->minus_time_interval("one") } qr/Bad format/, 'minus_time_interval("one") is not a mind-reader..';
+
+is($datetime1->get_nth_day_of_week(3, 3)->day_of_month, 21, 'nth_day_of_week(3, 3) is correct');
```
Also, as long as I am here, you will want the two arguments here to be distinct.  With `(3,3)` it would be possible for someone to switch the argument order in the code and still have passing tests even though the behavior would be wrong. 

The tests also act as a sort of runnable documentation, so we should be able to look at the test and know how the call works.

mahdix replied:
By using day of week names instead of numbers, this is solved, I believe.

Commit:
Create a single source of day name knowledge.

Admittedly, these seem very unlikely to change, but it's nice not to
wonder where each piece is done.  Also, this will all happen now on
package compliation, so each method call will be very slightly faster.

Also, completes the rename everywhere.

Commit:
Merge pull request #1 from mwmiller/mm/nth_dow

Create a single source of day name knowledge.

mwmiller commented on lib/Date/Utility.pm:
```diff
@@ -931,6 +925,32 @@ sub truncate_to_day {
+
+    my $dow = $day_abbrev_to_num{lc $dow_abb} or croak 'Invalid day of week. We got [' . $dow_abb . ']';
+
+    $dow %= 7;               # accept both 0 and 7 as Sunday
```
This line is no longer necessary; we've set the number for them.

mahdix replied:
Removed.

mwmiller commented on t/object_methods.t:
```diff
@@ -85,3 +85,11 @@ is($datetime3->minus_time_interval('1d')->is_same_as($datetime2),  1,          '
 throws_ok { $datetime3->minus_time_interval("one") } qr/Bad format/, 'minus_time_interval("one") is not a mind-reader..';
+
+subtest 'move_to_nth_dow' => sub {
+    is($datetime1->move_to_nth_dow(3, 'Wed')->day_of_month, 21, 'Third Wednesday of Dec 2012 is the 21st');
```
Looking at a calendar, this is not true! The 21st is a Friday.  I will look into what is happening here.

mwmiller replied:
Description is wrong because I am an idiot.  `datetime1` is 2011, not 2012!

mahdix replied:
I changed the comment :-)

Commit:
fix

mwmiller commented:
:+1:

torsten-deriv commented on lib/Date/Utility.pm:
```diff
@@ -592,20 +590,17 @@ Returns the name of the current day. Example: Sunday
+my %day_abbrev_to_num;
+foreach my $day_num (0 .. $#day_names) {
+    $day_abbrev_to_num{lc substr($day_names[$day_num], 0, 3)} = $day_num;
+}
```
What you can do is this:

``` perl
my %day_abbrev_to_num = map {
    my $day = $day_names[$_];
    (lc($day) => $_, lc(substr $day, 0, 3) =>$_, $_ => $_);
} 0..$#day_names;
```

That way you can pass full day names, abbreviated day names and day numbers (0..6) to the function below.

mahdix replied:
Matt has implemented this. I will merge his PR into mine.

torsten-deriv commented on lib/Date/Utility.pm:
```diff
@@ -931,6 +925,31 @@ sub truncate_to_day {
+
+    return unless $result->month == $self->month;
+    return $result;
+}
```
Perhaps, now that this function is part of this module it should also use the utilities provided by this module. Meaning no `Time::Local`, no `timegm` and no `gmtime`

mahdix replied:
Done in Matt's PR.

torsten-deriv commented on lib/Date/Utility.pm:
```diff
@@ -931,6 +925,31 @@ sub truncate_to_day {
+
+    my $result = Date::Utility->new($time + 24 * 3600 * $days_add);
+
+    return unless $result->month == $self->month;
```
Is this a safety measure against something like `$nth=19` or `$nth=-42`?

mahdix replied:
Yes.

torsten-deriv commented on lib/Date/Utility.pm:
```diff
@@ -931,6 +925,31 @@ sub truncate_to_day {
+=head2 move_to_nth_dow
+
+Returns undef or a Date::Utility object pointing to nth day of week of the current month.
+$dow is a three letter abbreviation of weekday names.
```
If one only reads the documentation without looking at the code, one cannot understand how to use the method.

mahdix replied:
Done by Matt.

Commit:
Expand available day of week representations.

Allow for full day names, thre letter abbrevations or numbers in the
range of 0 to 6.

Commit:
Improve documentation.

- Move section out of STATIC METHODS
- Add more documentation of how it is called

Commit:
Improve bounds checking.

Large values which cross years but happen to end up in the same month
should not return values.. because they don't make much sense.

Commit:
Improve code reuse.

Includes a new test to show that we can work with single digit months
and days in the format we're using.

Includes a bug fix for requesting Sunday (day 0)

Also requires that dates we supply exist with the change from
`gmtime_nocheck` to `gmtime_check`

Commit:
Use subtests for object method tests.

This should make understanding and maintaining this file a bit easier.

Commit:
Merge pull request #2 from mwmiller/mm/nth_dow

Mm/nth dow

torsten-deriv commented on t/object_methods.t:
```diff
@@ -2,86 +2,113 @@ use strict;
+    is($datetime1->move_to_nth_dow(3,   'Wed')->day_of_month,      21,    'Third Wednesday of Dec 2011 is the 21st');
+    is($datetime1->move_to_nth_dow(5,   'SuNdaY'),                 undef, '... and there is no 5th Sunday that year.');
+    is($datetime1->move_to_nth_dow(-50, 'Mon'),                    undef, '... nor a -50th Monday.');
+    is($datetime1->move_to_nth_dow(105, 'Tue'),                    undef, '... nor a 105th Tuesday.');
```
Good test cases would also be Feb 29 in 2000 (does not exist), in 2001 (does not exist) and in 2004 (does exist)

torsten-deriv replied:
What I tried to say is: It's perhaps also a good idea to test for cases that only just exceed the boundaries. `-50` and `105` are way off. But `0` is just out of bounds. Also, there is a 5th Wed in Sep 2015 but not in Aug 2015.

This kind of testing is not so much to make sure the current implementation works as expected but also against future refactoring.

mwmiller replied:
These particular tests were sort of as regression tests.  That is to say, I wrote failing tests against the old implementation and got them to pass with the new one.

I don't mind adding some more, but it's not clear to me where to add additional checks which will provide further assurance.  Showing a 5th Saturday with no 5th Sunday is trying to be that "edge check".

Famous last words, but it should be obvious that non-positive numbers are nonsensical, so adding more nonsense seems like overkill.

mwmiller replied:
> Good test cases would also be Feb 29 in 2000 (does not exist), in 2001 (does not exist) and in 2004 (does exist)

Indeed these would be some good checks, but I don't think they belong in this particular file.  Some failing cases in the construction make sense.

Commit:
Add leap year constructor tests.

Now using `timegm` instead of `timegm_nocheck` we should fail on clearly
invalid dates.

torsten-deriv commented on lib/Date/Utility.pm:
```diff
@@ -501,7 +499,7 @@ sub _parse_datetime_param {
     }
 
-    my $epoch = timegm_nocheck($second, $minute, $hour, $day, $month - 1, $year);
+    my $epoch = timegm($second, $minute, $hour, $day, $month - 1, $year);
```
This breaks stuff like:

``` perl
Date::Utility->new('2015-02-28')->plus_time_interval('1d');
```

please add tests to make sure this works.

torsten-deriv replied:
But simply changing it back to `timegm_nocheck` breaks one of the new tests. So, there is a bit more work to do.

mwmiller replied:
> This breaks stuff like:
> Date::Utility->new('2015-02-28')->plus_time_interval('1d');

It absolutely does not.  Why would think that it would?

mahdix replied:
It works correctly according to my tests. I also added a unit test for this.

torsten-deriv replied:
Ok, it works for me as well. No idea what I did wrong when I tried it.

torsten-deriv commented on lib/Date/Utility.pm:
```diff
@@ -840,6 +838,35 @@ sub months_ahead {
+
+    my $dow = $days_to_num{lc $dow_abb} // croak 'Invalid day of week. We got [' . $dow_abb . ']';
+
+    my $dow_first = ($self->day_of_month - $self->day_of_week) % 7;
```
this is wrong. The correct formula is:

``` perl
    my $dow_first = (1 + $self->day_of_week - $self->day_of_month) % 7;
```

mwmiller replied:
We add the one in the next line.  But the variable as a poor name for what it does.

torsten-deriv commented on lib/Date/Utility.pm:
```diff
@@ -840,6 +838,35 @@ sub months_ahead {
+    my $dom = ($dow + 7 - $dow_first) % 7 + ($nth - 1) * 7 + 1;
+
+    return try { Date::Utility->new(join '-', $self->year, $self->month, $dom) };
+}
```
Given the error above, I think this function needs better testing. The following piece is just an example. Ideally the `19` is also varied for all days of a month. Also 1000 days is just an arbitrary number.

``` perl
    my $d = Date::Utility->new(Date::Utility::today->year . '-01-01 12:00');
    my @dow;
    my $M = '';
    for (0..1000) {
        unless ($M eq $d->year.$d->month) {
            $M = $d->year.$d->month;
            @dow = ();
        }
        $dow[$d->day_of_week]++;
        is +Date::Utility->new(join '-', $d->year, $d->month, 19)
                         ->move_to_nth_dow($dow[$d->day_of_week], $d->day_of_week)
                         ->date,
           $d->date,
           "$dow[$d->day_of_week]. ".$d->full_day_name." of ".$d->month_as_string.' '.$d->year.' is '.$d->date;

        $d = $d->plus_time_interval('1d');
    }
```

mwmiller replied:
I did extensive testing (not committed in favor of fast running tests) and it was producing correct results.  I am not sure what you're demonstrating here.

mwmiller replied:
Ok, there appears to be a problem in months which start on Sunday (as day 0)... I'm looking into this.

mwmiller replied:
Ok, I ran your script with `2e3-1` replacing `1000` and a `for` loop of `1..15` replacing `19`:

```
ok 29986 - 4. Monday of Jun 2020 is 2020-06-22
ok 29987 - 4. Monday of Jun 2020 is 2020-06-22
ok 29988 - 4. Monday of Jun 2020 is 2020-06-22
ok 29989 - 4. Monday of Jun 2020 is 2020-06-22
ok 29990 - 4. Monday of Jun 2020 is 2020-06-22
ok 29991 - 4. Monday of Jun 2020 is 2020-06-22
ok 29992 - 4. Monday of Jun 2020 is 2020-06-22
ok 29993 - 4. Monday of Jun 2020 is 2020-06-22
ok 29994 - 4. Monday of Jun 2020 is 2020-06-22
ok 29995 - 4. Monday of Jun 2020 is 2020-06-22
ok 29996 - 4. Monday of Jun 2020 is 2020-06-22
ok 29997 - 4. Monday of Jun 2020 is 2020-06-22
ok 29998 - 4. Monday of Jun 2020 is 2020-06-22
ok 29999 - 4. Monday of Jun 2020 is 2020-06-22
ok 30000 - 4. Monday of Jun 2020 is 2020-06-22
1..30000
```

So, this seems to be working now.  The algorithm, as written, seems too complex because it matches the way I thought about solving it.  I am going to enshrine a smaller version of the test in the suite and work on a simplification.  Thanks for the note.. my extensive testing was along the wrong axis to find the issue!

mahdix replied:
@mwmiller You are right. For example for Mar 2015 (Starts with Sunday), move_to_nth_dow(1, 0) will return Mar-7 while it has to be Mar-1

mahdix replied:
I will fix this.

mwmiller replied:
@mm-binary This is fixed (with tests demonstrating correctness) in the pull against your fork.

mahdix replied:
Thanks.

Commit:
Correct day finding algorithm.

This looks more complex that I would like, but it maps to how I worked
through it on paper and I haven't found a simplication.

Commit:
Merge pull request #3 from mwmiller/mm/nth_dow

Mm/nth dow
